soroban-sdk-macros/src/lib.rs

```diff
@@ -95,6 +97,45 @@ pub fn contractimpl(_metadata: TokenStream, input: TokenStream) -> TokenStream {
     }
 }
 
+#[derive(Debug, FromMeta)]
+struct MetadataArgs {
+    key: String,
+    val: String,
+}
+
+#[proc_macro]
+pub fn metadata(metadata: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(metadata as AttributeArgs);
+    let args = match MetadataArgs::from_list(&args) {
+        Ok(v) => v,
+        Err(e) => return e.write_errors().into(),
+    };
+
+    let gen = {
+        //TODO: is key necessary?
+        let meta_v0 = ScMetaV0 {
+            key: args.key.clone().try_into().unwrap(),
+            val: args.val.try_into().unwrap(),
+        };
+        let meta_entry = ScMetaEntry::ScMetaV0(meta_v0);
+
+        let metadata_xdr = meta_entry.to_xdr().unwrap();
+        let metadata_xdr_lit = proc_macro2::Literal::byte_string(metadata_xdr.as_slice());
+        let metadata_xdr_len = metadata_xdr.len();
+
+        // TODO: Handle collisions on "metadata"?
+        quote! {
+            #[cfg_attr(target_family = "wasm", link_section = "metadata")]
```

Comment:
Can we call the section `contractmetav0`. We use the contract prefix on others and a v0 suffix on others. The v0 suffix is so that we can communicate about a v1 in comparison if we entirely abandon the format within.

(Note that we have env meta in a separate section, `contractenvmetav0`. They won't collide.)

Reply:
done, although there's still the collision issue to resolve here.